### DISCUSSION TITLE
What does parallelReduce do?

### DISCUSSION THREAD

**Discussion by TheChronicMonster** [reach-sh/reach-lang#920] [General]
`parallelReduce` is an abbreviation of a `while` loop combined with a [fork](https://docs.reach.sh/rsh/step/#rsh_fork).

`parallelReduce` allows a value to be repeatedly updated by racing participants until the `while` loop's condition is no longer true. 

https://docs.reach.sh/rsh/consensus/#parallelreduce--intuition

**Comment by hmzdot** [comment]
`parallelReduce` name comes from the fact that participants are trying to produce a new state from the current values in parallel. So while Alice is producing a new state, Bob might also be producing one. Parallel reduce is essentially a fork+case loop and what this really means is Reach first determines who tried to update the state first (fork) and depending on the winner of the race runs a specified logic (case)
So it's like reduce function in JavaScript but for multiple threads. In our case threads are different participants running the program

Note: In case you didn't know, reduction means taking a bunch of items and processing them recursively, so that result of the previous process become the input of the next one. As an example, if we were to reduce `[1, 2, 3]` list with `const add = (previous, next) => previous + next` function with starting value of 0, steps would be:
- prev: 0, next: 1, add(prev, next): 1
- prev: 1, next: 2, add(prev, next): 3
- prev: 3, next: 3, add(prev, next): 6
